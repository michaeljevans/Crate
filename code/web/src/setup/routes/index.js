// App Imports
import { APP_URL_API } from '../config/env'
import admin from './admin'
import home from './home'
import user from './user'
import product from './product'
import crate from './crate'

// Combined routes
// creating and exporting a new object with properties copied from each of the objects exported in the route files
export const routes = Object.assign(admin, home, user, product, crate)
console.log('ROUTES', routes)

// API Routes
// assigning the variable APP_URL_API (which has the value of http://localhost:8000 -- backend server), to the variable 'routeApi' and exporting it so the route is easily acccessible throughout app
export const routeApi = APP_URL_API

// Image
export const routeImage = APP_URL_API
