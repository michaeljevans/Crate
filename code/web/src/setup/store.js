// Imports
import { compose, combineReducers } from 'redux'
import { createStore, applyMiddleware } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'
import thunk from 'redux-thunk'

// App Imports
import common from '../modules/common/api/state'
import user from '../modules/user/api/state'
import * as product from '../modules/product/api/state'
import * as subscription from '../modules/subscription/api/state'
import * as crate from '../modules/crate/api/state'

// The combineReducers helper function turns an object whose values are different reducing functions into a single reducing function you can pass to createStore.


// App Reducer

// the combineReducers function turns the passed in reducing functions into a single reducing function. This single reducing function can be accessed by calling appReducer 

const appReducer = combineReducers({
  common,
  user,
  ...product,
  ...subscription,
  ...crate
})

// Root Reducer
// using an if statement to determine the value(?)/ what's returned(?) in the rootReducer
// if the action type is not 'RESET' then the rootReducer will invoke the appReducer function
// if the action is 'RESET' then the rootReducer will assign the value of state as undefined

export const rootReducer = (state, action) => {
  if (action.type === 'RESET') {
    state = undefined
  }

  return appReducer(state, action)
}

// Assigning a value to initial state if window is not undefined
// window._INITIAL_STATE_ is assigned value of "JSON.stringify(initialState)" in view.js, where initialState's default value is an empty object 
// then uses the delete keyword to delete the window.__INITIAL_STATE__ object
let initialState
if (typeof window !== 'undefined') {
  initialState = window.__INITIAL_STATE__
  console.log('initial state', initialState)
  delete window.__INITIAL_STATE__
}

// Store
export const store = createStore(
  rootReducer,
  initialState,
  composeWithDevTools(
    applyMiddleware(thunk),
  )
) 