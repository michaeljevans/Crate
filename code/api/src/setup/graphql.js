// Imports
import graphqlHTTP from 'express-graphql'

// App Imports
import serverConfig from '../config/server.json'
import authentication from './authentication'
import schema from './schema'

// Setup GraphQL
export default function (server) {
  console.info('SETUP - GraphQL...')

  server.use(authentication)

  // API (GraphQL on route `/`)
  // defining the single endpoint 
  // imports schema and tells server to use the schema
  // schema is a root directory(?) for queries (get/read data) and mutations (other API calls that update/mutate data)
  server.use(serverConfig.graphql.endpoint, graphqlHTTP(request => ({
    schema,
    graphiql: serverConfig.graphql.ide,
    pretty: serverConfig.graphql.pretty,
    context: {
      auth: {
        user: request.user,
        isAuthenticated: request.user && request.user.id > 0
      }
    }
  })))
}
