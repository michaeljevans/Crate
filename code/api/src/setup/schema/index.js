// Imports
import { GraphQLSchema } from 'graphql'

// App Imports
// These files have imported all queries and mutations from migration tables
import query from './queries'
import mutation from './mutations'

// Schema
const schema = new GraphQLSchema({
  query,
  mutation
})

export default schema
