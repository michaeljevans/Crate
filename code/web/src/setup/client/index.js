// Imports
import React from 'react'
import { hydrate } from 'react-dom'
import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux'

// App Imports
import { store } from '../../setup/store'
import { setUser, loginSetUserLocalStorageAndCookie } from '../../modules/user/api/actions'
import ScrollToTop from '../../modules/common/ScrollToTop'
import App from './App'

// User Authentication
// if user logs in there is a token to get from local storage? YES!!
// In addition to a users details, a token is returned from the resolver method (login) on the backend. 
// When a user logs in successfully the token and user are added to local storage
const token = window.localStorage.getItem('token')
if (token && token !== 'undefined' && token !== '') {
  // console.log('Token:', token)
  const user = JSON.parse(window.localStorage.getItem('user'))
  if (user) {
    // Dispatch action through setUser method(?) 
    // setUser returns SET_USER action with payload of the user object, which adds the user to store
    // is this how a user remains logged in even if they close the window?
    store.dispatch(setUser(token, user))
    // console.log('User:', user)

    loginSetUserLocalStorageAndCookie(token, user)
  }
}

// Client App
// wrapping the APP with <Provider > connects the APP to Redux store
const Client = () => (
  <Provider store={store} key="provider">
    <Router>
      <ScrollToTop>
        <App/>
      </ScrollToTop>
    </Router>
  </Provider>
)

// Mount client app
/* when window loads, render this code inside the 'app' html element */
window.onload = () => {
  hydrate(
    <Client/>,
    document.getElementById('app')
  )
}
