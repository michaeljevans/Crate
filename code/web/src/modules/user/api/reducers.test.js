import { rootReducer } from '../../../setup/store'
import { initialState } from '../../../setup/store'

describe('User Reducer', () => {

  it('Should return the initial state for user', () => {
    expect(rootReducer(undefined, {})).toMatchSnapshot()
  })

  it('Should handle a login request', () => {
    expect(
      rootReducer(initialState,
        {
          type: 'AUTH/LOGIN_REQUEST',
          isLoading: true
        }
      )
    ).toMatchSnapshot()
  })

  it('Should set a new user', () => {
    expect(
      rootReducer(initialState,
        {
          type: 'AUTH/SET_USER',
          user: {
            email: 'user@crate.com',
            name: 'The User',
            role: 'USER'
          }
        }
      )
    ).toMatchSnapshot()
  })

  it('Should handle a login response', () => {
    expect(
      rootReducer(initialState,
        {
          type: 'AUTH/LOGIN_RESPONSE',
          isLoading: false
        }
      )
    ).toMatchSnapshot()
  })
})