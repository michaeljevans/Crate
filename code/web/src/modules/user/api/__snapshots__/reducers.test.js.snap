// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Reducer Should handle a login request 1`] = `
Object {
  "common": Object {
    "message": Object {
      "open": false,
      "text": null,
    },
  },
  "crate": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "crates": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "product": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "products": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "productsRelated": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
    "productId": 0,
  },
  "subscription": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "subscriptions": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "subscriptionsByUser": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "user": Object {
    "details": null,
    "error": null,
    "isAuthenticated": false,
    "isLoading": true,
  },
}
`;

exports[`User Reducer Should handle a login response 1`] = `
Object {
  "common": Object {
    "message": Object {
      "open": false,
      "text": null,
    },
  },
  "crate": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "crates": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "product": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "products": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "productsRelated": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
    "productId": 0,
  },
  "subscription": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "subscriptions": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "subscriptionsByUser": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "user": Object {
    "details": null,
    "error": undefined,
    "isAuthenticated": false,
    "isLoading": false,
  },
}
`;

exports[`User Reducer Should return the initial state for user 1`] = `
Object {
  "common": Object {
    "message": Object {
      "open": false,
      "text": null,
    },
  },
  "crate": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "crates": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "product": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "products": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "productsRelated": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
    "productId": 0,
  },
  "subscription": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "subscriptions": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "subscriptionsByUser": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "user": Object {
    "details": null,
    "error": null,
    "isAuthenticated": false,
    "isLoading": false,
  },
}
`;

exports[`User Reducer Should set a new user 1`] = `
Object {
  "common": Object {
    "message": Object {
      "open": false,
      "text": null,
    },
  },
  "crate": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "crates": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "product": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "products": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "productsRelated": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
    "productId": 0,
  },
  "subscription": Object {
    "error": null,
    "isLoading": false,
    "item": Object {},
  },
  "subscriptions": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "subscriptionsByUser": Object {
    "error": null,
    "isLoading": false,
    "list": Array [],
  },
  "user": Object {
    "details": Object {
      "email": "user@crate.com",
      "name": "The User",
      "role": "USER",
    },
    "error": null,
    "isAuthenticated": true,
    "isLoading": false,
    "style": null,
  },
}
`;
